src/driver/buffers.rs

```diff
@@ -0,0 +1,132 @@
+use crate::runtime::CONTEXT;
+
+use libc::{iovec, UIO_MAXIOV};
+use std::cell::RefCell;
+use std::cmp;
+use std::io;
+use std::mem;
+use std::ptr;
+use std::rc::Rc;
+use std::slice;
+
+pub(crate) struct Buffers {
+    raw_vecs: ptr::NonNull<iovec>,
+    buf_states: Vec<BufState>,
+    orig_cap: usize,
+}
+
+enum BufState {
+    Free { init_len: usize },
+    CheckedOut,
+}
+
+impl Buffers {
+    pub(crate) fn new(bufs: impl Iterator<Item = Vec<u8>>) -> Self {
```

Comment:
Is this Item of type Vec<u8> the only and best choice? Is there a more primitive type that would serve the same purpose? Maybe its a good choice because the std library comes with so many ways to unsafely create a Vec from a larger allocation?

Reply:
I used `Vec<u8>` as a safe representation of an allocated buffer. I'm not aware of another primitive type to provide the semantics of memory allocation and management of initialized vs. uninitialized data. It could be made more convenient if the initializing iterator items were generic and bound with `Into<Vec<u8>>`, but I do not want to hide the act of providing a number of allocated buffers.

As mentioned above, the registry API can be easily made generic around the allocator the same way `Vec` is.

Otherwise, I have not yet thought of making the fixed buffers API more generic. I think the primary use case here is to allocate one or several flat byte buffers, register them with the kernel, and then use the buffers repeatedly with I/O operations, filling or consuming buffer data as needed by the application with its program-friendly memory structures living elsewhere.